buckets/redis/bucket.go

```diff
@@ -59,28 +59,65 @@ func NewBucketFactory(redisOpts *redis.Options, connectionRetries int) quotaserv
 }
 
 func (bf *bucketFactory) Init(cfg *pbconfig.ServiceConfig) {
+	logging.Printf("Initializing redis.bucketFactory for config version %v", cfg.Version)
 	bf.mu.Lock()
 	defer bf.mu.Unlock()
 
 	bf.cfg = cfg
 
 	if bf.client == nil {
+		start := time.Now()
 		bf.connectToRedisLocked()
+		logging.Printf("Re-established Redis connections in %v", time.Since(start))
 	}
 
-	// Clean out all buckets, start afresh.
+	// Validate contents in Redis
+	start := time.Now()
+	// Check if a flush intent is in place for this config version. If it is, it means that another node has
+	// already flushed Redis and this node doesn't have to.
+	intent := bf.client.Get("FlushIntent")
+	version, err := intent.Int64()
+	if err != nil {
+		if err.Error() != "redis: nil" {
+			// An actual problem; not just a missing flush intent
+			logging.Fatalf("Error response from Redis: %v", err)
+		}
+
+		// No version has been established.
+		bf.setIntentAndFlush(cfg.Version)
+
+	} else {
+		logging.Printf("Flush intent detected for version %v", version)
+		if int32(version) < cfg.Version {
+			bf.setIntentAndFlush(cfg.Version)
+		} else {
+			logging.Printf("No need to flush since a flush for version %v has already taken place", version)
+		}
+	}
+	logging.Printf("Verified Redis (including any flushes, if necessary) in %v", time.Since(start))
+}
+
+func (bf *bucketFactory) setIntentAndFlush(version int32) {
+	start := time.Now()
+	// Store flush intent; intent expires after 1 minute. It doesn't really matter if a new flush intent is
+	// stored before this intent expires, since the config version that triggered the intent is persisted.
+	bf.client.Set("FlushIntent", version, 1*time.Minute)
+
 	bf.client.FlushDb()
+	// Re-set flush-intent, since previous intent would have been removed
+	bf.client.Set("FlushIntent", version, 1*time.Minute)
```

Comment:
I wonder we actually need to expire this. We can just persist forever to avoid any further `FlushDb` to the same (or lower) version?